File: mutate4py/src/mutate4py/analysis.py
from __future__ import annotations

import ast
import hashlib
from dataclasses import dataclass
from pathlib import Path

from mutate4py.model import MutationScope, MutationSite, SourceAnalysis


class MutationCatalog:
    def discover(self, files: list[str]) -> list[MutationSite]:
        sites: list[MutationSite] = []
        for file_path in files:
            sites.extend(self.analyze(file_path).sites)
        sites.sort(key=lambda site: (site.file, site.start))
        return sites

    def analyze(self, file_path: str) -> SourceAnalysis:
        source = Path(file_path).read_text(encoding="utf-8")
        tree = ast.parse(source, filename=file_path)
        offsets = _line_offsets(source)
        visitor = _MutationVisitor(file_path, source, offsets)
        visitor.visit(tree)
        scopes = sorted(visitor.scopes.values(), key=lambda scope: scope.id)
        if not scopes:
            scopes = [
                MutationScope(
                    id=f"file:{Path(file_path).name}",
                    kind="file",
                    start_line=1,
                    end_line=max(1, source.count("\n") + 1),
                    semantic_hash=_hash_text(source),
                )
            ]
        module_hash = _hash_text(
            "\n".join(f"{scope.id}|{scope.semantic_hash}" for scope in scopes)
        )
        return SourceAnalysis(
            source=source, sites=visitor.sites, scopes=scopes, module_hash=module_hash
        )


@dataclass(frozen=True)
class _ScopeRef:
    id: str
    kind: str
    start_line: int
    end_line: int


class _MutationVisitor(ast.NodeVisitor):
    def __init__(self, file_path: str, source: str, offsets: list[int]) -> None:
        self.file_path = file_path
        self.source = source
        self.offsets = offsets
        self.sites: list[MutationSite] = []
        self.scopes: dict[str, MutationScope] = {}
        self.scope_stack: list[_ScopeRef] = []

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        self._visit_function(node, "function")

    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
        self._visit_function(node, "function")

    def visit_Constant(self, node: ast.Constant) -> None:
        current = self._current_scope(node)
        if isinstance(node.value, bool):
            original = "True" if node.value else "False"
            replacement = "False" if node.value else "True"
            self._add_site(
                node,
                original,
                replacement,
                f"replace {original} with {replacement}",
                current,
            )
        elif isinstance(node.value, int) and node.value in (0, 1):
            original = str(node.value)
            replacement = "1" if node.value == 0 else "0"
            self._add_site(
                node,
                original,
                replacement,
                f"replace {original} with {replacement}",
                current,
            )
        self.generic_visit(node)

    def visit_Compare(self, node: ast.Compare) -> None:
        current = self._current_scope(node)
        if len(node.ops) == 1 and len(node.comparators) == 1:
            replacement = _compare_replacement(node.ops[0])
            if replacement is not None:
                left_end = self._node_offset(node.left, end=True)
                right_start = self._node_offset(node.comparators[0], end=False)
                operator = self.source[left_end:right_start].strip()
                start = self.source.find(operator, left_end, right_start)
                if start >= 0:
                    self.sites.append(
                        MutationSite(
                            file=self.file_path,
                            line=node.lineno,
                            start=start,
                            end=start + len(operator),
                            original_text=operator,
                            replacement_text=replacement,
                            description=f"replace {operator} with {replacement}",
                            scope_id=current.id,
                            scope_kind=current.kind,
                            scope_start_line=current.start_line,
                            scope_end_line=current.end_line,
                        )
                    )
        self.generic_visit(node)

    def visit_BinOp(self, node: ast.BinOp) -> None:
        current = self._current_scope(node)
        replacement = _binop_replacement(node.op)
        if replacement is not None:
            left_end = self._node_offset(node.left, end=True)
            right_start = self._node_offset(node.right, end=False)
            operator = self.source[left_end:right_start].strip()
            start = self.source.find(operator, left_end, right_start)
            if start >= 0:
                self.sites.append(
                    MutationSite(
                        file=self.file_path,
                        line=node.lineno,
                        start=start,
                        end=start + len(operator),
                        original_text=operator,
                        replacement_text=replacement,
                        description=f"replace {operator} with {replacement}",
                        scope_id=current.id,
                        scope_kind=current.kind,
                        scope_start_line=current.start_line,
                        scope_end_line=current.end_line,
                    )
                )
        self.generic_visit(node)

    def visit_BoolOp(self, node: ast.BoolOp) -> None:
        current = self._current_scope(node)
        replacement = _boolop_replacement(node.op)
        if replacement is not None and len(node.values) >= 2:
            left_end = self._node_offset(node.values[0], end=True)
            right_start = self._node_offset(node.values[1], end=False)
            operator = self.source[left_end:right_start].strip()
            start = self.source.find(operator, left_end, right_start)
            if start >= 0:
                self.sites.append(
                    MutationSite(
                        file=self.file_path,
                        line=node.lineno,
                        start=start,
                        end=start + len(operator),
                        original_text=operator,
                        replacement_text=replacement,
                        description=f"replace {operator} with {replacement}",
                        scope_id=current.id,
                        scope_kind=current.kind,
                        scope_start_line=current.start_line,
                        scope_end_line=current.end_line,
                    )
                )
        self.generic_visit(node)

    def visit_UnaryOp(self, node: ast.UnaryOp) -> None:
        current = self._current_scope(node)
        operator = None
        if isinstance(node.op, ast.Not):
            operator = "not"
        elif isinstance(node.op, ast.USub):
            operator = "-"
        if operator is not None:
            start = self._node_offset(node, end=False)
            operand_start = self._node_offset(node.operand, end=False)
            slice_text = self.source[start:operand_start]
            index = slice_text.find(operator)
            if index >= 0:
                operator_start = start + index
                self.sites.append(
                    MutationSite(
                        file=self.file_path,
                        line=node.lineno,
                        start=operator_start,
                        end=operator_start + len(operator),
                        original_text=operator,
                        replacement_text="",
                        description=f"replace {operator} with removed {operator}",
                        scope_id=current.id,
                        scope_kind=current.kind,
                        scope_start_line=current.start_line,
                        scope_end_line=current.end_line,
                    )
                )
        self.generic_visit(node)

    def _visit_function(self, node: ast.AST, kind: str) -> None:
        name = node.name  # type: ignore[attr-defined]
        start_line = node.lineno  # type: ignore[attr-defined]
        end_line = node.end_lineno  # type: ignore[attr-defined]
        scope_id = f"{kind}:{name}:{start_line}"
        start = self._node_offset(node, end=False)
        end = self._node_offset(node, end=True)
        self.scopes[scope_id] = MutationScope(
            id=scope_id,
            kind=kind,
            start_line=start_line,
            end_line=end_line,
            semantic_hash=_hash_text(self.source[start:end]),
        )
        self.scope_stack.append(_ScopeRef(scope_id, kind, start_line, end_line))
        self.generic_visit(node)
        self.scope_stack.pop()

    def _add_site(
        self,
        node: ast.AST,
        original: str,
        replacement: str,
        description: str,
        current: _ScopeRef,
    ) -> None:
        start = self._node_offset(node, end=False)
        end = self._node_offset(node, end=True)
        self.sites.append(
            MutationSite(
                file=self.file_path,
                line=getattr(node, "lineno", 1),
                start=start,
                end=end,
                original_text=original,
                replacement_text=replacement,
                description=description,
                scope_id=current.id,
                scope_kind=current.kind,
                scope_start_line=current.start_line,
                scope_end_line=current.end_line,
            )
        )

    def _current_scope(self, node: ast.AST) -> _ScopeRef:
        if self.scope_stack:
            return self.scope_stack[-1]
        line = getattr(node, "lineno", 1)
        end_line = getattr(node, "end_lineno", line)
        return _ScopeRef("file:" + Path(self.file_path).name, "file", line, end_line)

    def _node_offset(self, node: ast.AST, *, end: bool) -> int:
        line = (
            getattr(node, "end_lineno", getattr(node, "lineno", 1))
            if end
            else getattr(node, "lineno", 1)
        ) - 1
        column = (
            getattr(node, "end_col_offset", getattr(node, "col_offset", 0))
            if end
            else getattr(node, "col_offset", 0)
        )
        return self.offsets[line] + column


def _line_offsets(source: str) -> list[int]:
    offsets = [0]
    for index, char in enumerate(source):
        if char == "\n":
            offsets.append(index + 1)
    return offsets


def _hash_text(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _compare_replacement(node: ast.AST) -> str | None:
    if isinstance(node, ast.Eq):
        return "!="
    if isinstance(node, ast.NotEq):
        return "=="
    if isinstance(node, ast.Gt):
        return ">="
    if isinstance(node, ast.GtE):
        return ">"
    if isinstance(node, ast.Lt):
        return "<="
    if isinstance(node, ast.LtE):
        return "<"
    return None


def _binop_replacement(node: ast.AST) -> str | None:
    if isinstance(node, ast.Add):
        return "-"
    if isinstance(node, ast.Sub):
        return "+"
    if isinstance(node, ast.Mult):
        return "/"
    if isinstance(node, ast.Div):
        return "*"
    return None


def _boolop_replacement(node: ast.AST) -> str | None:
    if isinstance(node, ast.And):
        return "or"
    if isinstance(node, ast.Or):
        return "and"
    return None

File: mutate4rs/src/analysis.rs
use std::{
    collections::BTreeMap,
    fs,
    path::{Path, PathBuf},
};

use proc_macro2::LineColumn;
use syn::{
    spanned::Spanned, visit::Visit, BinOp, ExprBinary, ExprLit, ExprUnary, File, ItemFn, Lit, UnOp,
};

use crate::model::{MutationScope, MutationSite, SourceAnalysis};

pub struct MutationCatalog;

impl MutationCatalog {
    pub fn discover(&self, files: &[PathBuf]) -> Result<Vec<MutationSite>, String> {
        let mut sites = Vec::new();
        for file in files {
            sites.extend(self.analyze(file)?.sites);
        }
        sites.sort_by(|left, right| {
            left.file
                .cmp(&right.file)
                .then(left.start.cmp(&right.start))
        });
        Ok(sites)
    }

    pub fn analyze(&self, file: &Path) -> Result<SourceAnalysis, String> {
        let source = fs::read_to_string(file).map_err(|err| err.to_string())?;
        let parsed: File = syn::parse_file(&source).map_err(|err| err.to_string())?;
        let line_offsets = line_offsets(&source);
        let (mut scopes, mut sites) = {
            let mut visitor = MutationVisitor::new(file, &source, line_offsets);
            visitor.visit_file(&parsed);
            (
                visitor.scopes.into_values().collect::<Vec<_>>(),
                visitor.sites,
            )
        };
        scopes.sort_by(|left, right| left.id.cmp(&right.id));
        if scopes.is_empty() {
            scopes.push(MutationScope {
                id: format!(
                    "file:{}",
                    file.file_name()
                        .and_then(|name| name.to_str())
                        .unwrap_or("unknown.rs")
                ),
                kind: "file".to_string(),
                start_line: 1,
                end_line: source.lines().count().max(1),
                semantic_hash: hash_text(&source),
            });
        }
        sites.sort_by(|left, right| left.start.cmp(&right.start));
        let module_hash = hash_text(
            &scopes
                .iter()
                .map(|scope| format!("{}|{}", scope.id, scope.semantic_hash))
                .collect::<Vec<_>>()
                .join("\n"),
        );

        Ok(SourceAnalysis {
            source,
            sites,
            scopes,
            module_hash,
        })
    }
}

#[derive(Clone)]
struct ScopeRef {
    id: String,
    kind: String,
    start_line: usize,
    end_line: usize,
}

struct MutationVisitor<'a> {
    file: &'a Path,
    source: &'a str,
    line_offsets: Vec<usize>,
    sites: Vec<MutationSite>,
    scopes: BTreeMap<String, MutationScope>,
    scope_stack: Vec<ScopeRef>,
}

impl<'a> MutationVisitor<'a> {
    fn new(file: &'a Path, source: &'a str, line_offsets: Vec<usize>) -> Self {
        Self {
            file,
            source,
            line_offsets,
            sites: Vec::new(),
            scopes: BTreeMap::new(),
            scope_stack: Vec::new(),
        }
    }

    fn current_scope(&self, line: usize) -> ScopeRef {
        self.scope_stack
            .last()
            .cloned()
            .unwrap_or_else(|| ScopeRef {
                id: format!(
                    "file:{}",
                    self.file
                        .file_name()
                        .and_then(|name| name.to_str())
                        .unwrap_or("unknown.rs")
                ),
                kind: "file".to_string(),
                start_line: line,
                end_line: line,
            })
    }

    fn offset(&self, position: LineColumn) -> usize {
        self.line_offsets[position.line - 1] + position.column
    }

    fn add_site(
        &mut self,
        line: usize,
        start: usize,
        end: usize,
        original: &str,
        replacement: &str,
        description: String,
    ) {
        let scope = self.current_scope(line);
        self.sites.push(MutationSite {
            file: self.file.to_string_lossy().replace('\\', "/"),
            line,
            start,
            end,
            original_text: original.to_string(),
            replacement_text: replacement.to_string(),
            description,
            scope_id: scope.id,
            scope_kind: scope.kind,
            scope_start_line: scope.start_line,
            scope_end_line: scope.end_line,
        });
    }
}

impl<'ast, 'a> Visit<'ast> for MutationVisitor<'a> {
    fn visit_item_fn(&mut self, node: &'ast ItemFn) {
        let start = node.sig.ident.span().start().line;
        let end = node.block.brace_token.span.close().end().line;
        let id = format!("function:{}:{}", node.sig.ident, start);
        let span = node.span();
        let snippet = &self.source[self.offset(span.start())..self.offset(span.end())];
        self.scopes.insert(
            id.clone(),
            MutationScope {
                id: id.clone(),
                kind: "function".to_string(),
                start_line: start,
                end_line: end,
                semantic_hash: hash_text(snippet),
            },
        );
        self.scope_stack.push(ScopeRef {
            id,
            kind: "function".to_string(),
            start_line: start,
            end_line: end,
        });
        syn::visit::visit_item_fn(self, node);
        self.scope_stack.pop();
    }

    fn visit_expr_lit(&mut self, node: &'ast ExprLit) {
        match &node.lit {
            Lit::Bool(boolean) => {
                let span = boolean.span();
                let start = self.offset(span.start());
                let end = self.offset(span.end());
                let original = if boolean.value { "true" } else { "false" };
                let replacement = if boolean.value { "false" } else { "true" };
                self.add_site(
                    span.start().line,
                    start,
                    end,
                    original,
                    replacement,
                    format!("replace {original} with {replacement}"),
                );
            }
            Lit::Int(int) if int.base10_digits() == "0" || int.base10_digits() == "1" => {
                let span = int.span();
                let start = self.offset(span.start());
                let end = self.offset(span.end());
                let original = int.base10_digits();
                let replacement = if original == "0" { "1" } else { "0" };
                self.add_site(
                    span.start().line,
                    start,
                    end,
                    original,
                    replacement,
                    format!("replace {original} with {replacement}"),
                );
            }
            _ => {}
        }
        syn::visit::visit_expr_lit(self, node);
    }

    fn visit_expr_binary(&mut self, node: &'ast ExprBinary) {
        if let Some(replacement) = binary_replacement(&node.op) {
            let left_end = self.offset(node.left.span().end());
            let right_start = self.offset(node.right.span().start());
            let between = &self.source[left_end..right_start];
            let operator = between.trim();
            if !operator.is_empty() {
                if let Some(relative) = between.find(operator) {
                    let start = left_end + relative;
                    let end = start + operator.len();
                    self.add_site(
                        node.span().start().line,
                        start,
                        end,
                        operator,
                        replacement,
                        format!("replace {operator} with {replacement}"),
                    );
                }
            }
        }
        syn::visit::visit_expr_binary(self, node);
    }

    fn visit_expr_unary(&mut self, node: &'ast ExprUnary) {
        if let Some(operator) = unary_operator(&node.op) {
            let expr_start = self.offset(node.span().start());
            let operand_start = self.offset(node.expr.span().start());
            let between = &self.source[expr_start..operand_start];
            if let Some(relative) = between.find(operator) {
                let start = expr_start + relative;
                let end = start + operator.len();
                self.add_site(
                    node.span().start().line,
                    start,
                    end,
                    operator,
                    "",
                    format!("replace {operator} with removed {operator}"),
                );
            }
        }
        syn::visit::visit_expr_unary(self, node);
    }
}

fn line_offsets(source: &str) -> Vec<usize> {
    let mut offsets = vec![0];
    for (index, ch) in source.char_indices() {
        if ch == '\n' {
            offsets.push(index + 1);
        }
    }
    offsets
}

fn hash_text(text: &str) -> String {
    use std::hash::{Hash, Hasher};
    let mut hasher = std::collections::hash_map::DefaultHasher::new();
    text.hash(&mut hasher);
    format!("{:016x}", hasher.finish())
}

fn binary_replacement(op: &BinOp) -> Option<&'static str> {
    Some(match op {
        BinOp::Eq(_) => "!=",
        BinOp::Ne(_) => "==",
        BinOp::Gt(_) => ">=",
        BinOp::Ge(_) => ">",
        BinOp::Lt(_) => "<=",
        BinOp::Le(_) => "<",
        BinOp::Add(_) => "-",
        BinOp::Sub(_) => "+",
        BinOp::Mul(_) => "/",
        BinOp::Div(_) => "*",
        BinOp::And(_) => "||",
        BinOp::Or(_) => "&&",
        _ => return None,
    })
}

fn unary_operator(op: &UnOp) -> Option<&'static str> {
    Some(match op {
        UnOp::Not(_) => "!",
        UnOp::Neg(_) => "-",
        _ => return None,
    })
}
